components/feature/push/src/main/java/mozilla/components/feature/push/AutoPushFeature.kt

```diff
@@ -338,6 +286,22 @@ class AutoPushFeature(
         return (System.currentTimeMillis() - prefLastVerified) >= PERIODIC_INTERVAL_MILLISECONDS
     }
 
+    /**
+     * Observers that want to receive updates for new subscriptions and messages.
+     */
+    interface Observer {
+
+        /**
+         * A subscription for the scope is available.
+         */
+        fun onSubscriptionChanged(scope: String) = Unit
+
+        /**
+         * A messaged has been received for the [scope].
+         */
+        fun onMessageReceived(scope: String, message: ByteArray?) = Unit
```

Comment:
Looking at your FxA code, this API feels a bit like a foot-gun. I _think_ you already have a problem in one consumer, where it's not filtering messages by `scope` and just tries to process everything. How can we then expect other consumers not to repeat this error themselves?

Feels to me that you'd ideally have something like `observeMessages(scope: String, handler:...)`, and `observeAllMessages(handler:...)`. Currently you only have the latter (`register(...)`).

Reply:
I'm a bit confused by this since we wanted to remove the MessageBus and have consumers ensure they weren't invoking their code unnecessarily. Do we want to go back on this?

Reply:
Thinking about GV API is useful here - I'm assuming it's shaped something like "onMessageReceived(scope, message)", right? Since it needs to handle messages for arbitrary scopes, I'm not sure what else it can be. If that's the case, you're totally right about the shape of this!

Something to consider for the future is a perf improvement - if we _know_ that a message is received for the deterministic FxA scope (see my other comments about this), we know that we don't need to send it to GV, and it could go straight to our FxA consumer. And vice versa, no need to send GV-bound messages to FxA. This may imply an API level change (you need to somehow know which scopes observers care/don't care about), or you could do this by inspecting observers themselves (e.g. which class is the observer?). The latter does feel a bit like magic to me and seemingly violates the public API... but it's an easy way to achieve this improvement, if we need it.